### ISSUE TITLE
Feature-finalrej

### ISSUE THREAD

**Issue by georgebuzzell** [NDCLab/baseEEG#28]
Feature-finalrej
-loop through each channel. For a given channel, loop over all epochs for that channel and identify epochs for which that channel, for a given epoch, exceeds either the voltage threshold or spectral threshold. If it exceeds the threshold, reject the channel data for this channel/epoch.
-write to the annotations file which channel/epoch intersections were rejected

**Comment by F-said** [comment]
Before making a PR for [dev-feature-finalRej-farukh](/NDCLab/baseEEG/tree/dev-feature-finalRej-farukh) I just wanted to sanity-check my code by running a few lines by you @georgebuzzell 

To filter epochs I create this rejection dictionary that specifies the "maximum acceptable peak-to-peak amplitude" criteria for EEG data. To clarify my EEG knowledge, is this filter anywhere on the right track for "max voltage" or "spectral threshold"? 

```python
reject = dict(eeg=100e-6)
```

Afterwards, I literally loop through all channel names while dropping epochs specified by the events created in Jonhas's segment code:

```python
channels = raw.info.ch_names

for label in channels:
    epochs = mne.Epochs(raw, events, event_id=event_id, tmin=-0.2, tmax=0.5,
                picks=label, preload=True, reject=reject, reject_by_annotation=True)
        
    epochs.drop_bad()
```

And finally, I log the dropped epochs to a log file, perhaps a starting point for the annotations file?

```python
logging.info("The following channels were rejected due to exceeding either the voltage threshold or spectral threshold ")
logging.info(epochs.drop_log)
```

Please let me know what you think! The script in its entirety could be found at [finalrej.ipynb](https://github.com/NDCLab/baseEEG/blob/dev-feature-finalRej-farukh/base_eeg/finalrej.ipynb). Thanks.

**Comment by georgebuzzell** [comment]
@F-said This is fantastic! Your interpretations are all correct and this code is an excellent first stab here. In fact, from what I can tell your code is essentially the standard way of dealing with rejections. There is, however, a way to improve this further. 

Basically, what you want to do is modify your code to NOT reject all channels for a segment (i.e. remove the entire segment) if only a select few channels are bad for that segment **(please correct me if I am wrong and you are already doing this)**. Instead, you want to loop through channels and only reject a given channel for a given segment, not all channels. When we implemented something like this in matlab/eeglab, we had to do a hack/workaround to allow for rejecting individual channels for a segment, because this yields unequal numbers of channels across segments, and since eeglab/matlab stores data as matrices/arrays, you cannot have unequal channel numbers across segments. 

**Can you please confirm whether or not you are currently allowing for select channels to be rejected for a given segment but still retaining the remaining channels for that segment (preferred approach) vs rejecting all channels for a segment even if just one or a select few are bad (traditional approach)?**

Note: pending how mne is storing the epoch/segmented data (I have not looked into this yet) you might have to do a workaround to allow for channel-level rejection and unequal channels across epochs/segments. One option is to just write in zeros in place of rejected channels, to maintain the same dimensions across segments as a placeholder; then, in a later step (beyond the scope of this feature) we would interpolate those zeroed-out channels with the other remaining channels for the segment. Another option is to interpolate the missing channels on the fly, as part of this feature, as opposed to writing in zeros first; however, I slightly prefer the zero method. If you want to interpolate on the fly, then you definitely want to involve @Jonhas who is working on the interpolate feature. It is also possible that mne will allow for unequal channels across segments, in which case you don't need to worry about these workarounds.

**Comment by georgebuzzell** [comment]
@F-said please see https://github.com/NDCLab/baseEEG/issues/25 and 'Autoreject' reject as well!

**Comment by F-said** [comment]
@georgebuzzell One question: is writing to a .log file acceptable for the annotations file we will eventually be building or would something simpler like a .txt file suffice for annotating transformations to data? The .log will be annotating things like light warnings as well as our method outputs.

**Comment by georgebuzzell** [comment]
@F-said Good question!

For the annotations, I think it would be good to go with a .json file for consistency with BIDS. Did you see the updated roadmap that gives a brief description of the annotations file? That said, I do certainly see value in an ADDITIONAL .log file that writes all warnings, etc. I think that is a GREAT idea. do you see an issue with doing both? or, relatedly, do you think it is redundant to have both? The reasoning for both would be that the .log file is something that will be a mess of output, but the JSON file would have the key outputs of interest written.

I sort of think of the JSON as the standard output that people would look at, and the .log would be a "verbose" output with EVERYTHING. thoughts?

**Comment by georgebuzzell** [comment]
repost from @F-said : Got it, thanks for pointing me to the right resource. I definitely see value in recording verbose outputs and mne warnings to find bad/unexpected behavior. Would this also serve any purpose on the research side of things? Like if this code was accessed by other researchers, would they find value in seeing these verbose outputs?

**Comment by georgebuzzell** [comment]
@F-said yes, could certainly have research value, mainly in terms of catching a warning that maybe someone disregarded that caused unexpected data outputs. However, the primary research parameters of interest should all go into the output of the annotation .json

Basically, I imagine it like this: when we write up the results of a project, we have to write a "methods" section describing how we processed our data. It should be possible to take the input and output .json files and run a script on them that literally generate the methods section for you. This is not a common thing that is currently done, but some researchers are starting to do things like this. this is what we should do and this is one of the primary purposes of these input/output json files (of course, the input json actually controls the analysis as well, so it is for more than just reporting).

@yanbin-niu @Jonhas @DMRoberts @F-said @DavyNeat @SDOsmany Please all take note of my comments above regarding the purpose of the input/output .JSON files, as well as the additional usefulness of a .log file that @F-said suggests

**Comment by F-said** [comment]
@SDOsmany `autoreject` would work great here since it specifically targets epochs. Using the same code as discussed in #25 , we could find the bad epochs without dropping or interpolation:

```
ar.fit(epochs)
reject_log = ar.get_reject_log(epochs)
```

We then would would want to use the resultant boolean list or rejection dictionary to somehow zero out the bad epochs, as opposed to getting rid of them.

**Comment by SDOsmany** [comment]
@F-said thanks for the mention, I agree with using auto-reject for this issue, but can you be more specific as to what you mean by zero out the bad epochs?

**Comment by georgebuzzell** [comment]
@F-said Actually, this "finalRej" feature is a processing step near the end of the pipeline, right before we do interpolations, so, it is actually fine to use the typical autoreject functionality and let it take care of the interpolation for you as well! 

My understanding is that autoreject will identify bad channels, at the epoch level, reject them, and interpolate them. That is exactly what we want ultimately, so lets just use autoreject as it currently works, unless i misunderstand?

of note, when this finalrej feature runs, it will run at a stage in the pipeline where other channels have already been completely removed (from the earlier rejbadchans feature). what i am not sure, is if autoreject will also interpolate those channels that are completely missing, even before autorreject runs. If so, great, this solves all our problems and actually makes the next step (feature-interpolate) obsolate. @Jonhas please take note of this and communicate with @F-said here, in case that is the case. If so, no need to finish out the interpolate feature (although your initial work was already very helpful in terms of dealing with montages, so not a waste at all!).

However, if autoreject only interpolates the channesl that it finds as bad (not ones already removed by feature-rejBadChans) then we would still want the interpolation code that @Jonhas is working on to run, after autoreject, to further interpolate the channels that are completely missing (not just the ones that autoreject removed at the epoch level).

In sum, @F-said please: 1)  allow autoreject to interpolate, and no need to "zero out" as I previously suggested; 2) determine if autoreject only interpolates channels that it rejects at the epoch level, or if, it will also interpolate extra bad channels you remove before even running autoreject; 3) pending whether autoreject interpolates only at the epoch level channels it rejects, or if interpolates all channels, please coordinate with @Jonhas to deterimine if feature-interp is still needed or not.

Let me know if you have any questions! Thank you!!

**Comment by SDOsmany** [comment]
@georgebuzzell i think auto reject does interpolation aswell because when you runt ar.fit(epochs)
it outputs :

Repairing epochs: 100%|██████████| 1868/1868 [00:03<00:00, 482.23it/s]

n_interp:   0%|          | 0/3 [00:00<?, ?it/s]

it also outputs :

Fold:  80%|████████  | 8/10 [00:50<00:12,  6.36s/it]


Fold:  90%|█████████ | 9/10 [00:56<00:06,  6.36s/it]

but I'm not sure what the Fold represents.

**Comment by georgebuzzell** [comment]
@SDOsmany yes, autoreject definitely does interpolate the channels that it identifies as bad, for a given segment of data (epoch). However, we will also have removed additional channels for all segments (actually removed back when the channels were still continuous), and, we want to know if autoreject will also interpolate those by default or not. Perhaps you can take a look at the code or documentation to try to suss out whether auto-reject only interpolates the channels that you pass to it (i.e. does not reject the ones we completely removed before autoreject) or if it will interpolate all channels, for example, by referencing a montage and detecting channels missing from that montage, whether autoreject removed them, or we previously removed them. Does that make sense?

@F-said

**Comment by SDOsmany** [comment]
@georgebuzzell when you say removed the channels back when they were continuous you mean marked as bad or completely removed from the data? 

auto-reject says that the picks parameter is set to none by default. It also says that by default autoreject ignores the channels marked as bad in epochs.info['bads']

picks : ndarray, shape(n_channels) | None
    The channels to be considered for autoreject. If None, defaults
    to data channels {'meg', 'eeg'}, which will lead fitting and combining
    autoreject solutions across these channel types. Note that, if picks is
    None, autoreject ignores channels marked bad in epochs.info['bads'].

Does this answer the question?

**Comment by georgebuzzell** [comment]
@SDOsmany this is helpful! It is clear from this that autoreject does not run on channels marked as bad. However, a remaining question, is what happens to those channels marked as bad when autoreject runs interpolation. Does it just ignore those during that step, or interpolate those as well? Perhaps you can test this by marking a couple channels as bad, running autoreject, and then seeing if they are interpolated or not?

PS: I was referring to EITHER completely removing channels or marking them as bad, if that makes sense.

**Comment by SDOsmany** [comment]
@F-said I'm pretty sure auto-reject ignores the channels marked as bad and it does not interpolate or repair them. 

the reject_log has a parameter called labels. You can call it by doing `reject_log = ar.get_reject_log(epochs)` and then `reject_log.labels`

in the documentation, it says: 
`labels: array, shape (n_epochs, n_channels)
    It contains integers that encode if a channel in a given
    epoch is good (value 0), bad (1), or bad and interpolated (2).`
    
    I built a dataframe from the `reject_log.labels` and it shows the 10 channels and the first and sixth channel have values of NaN.  These are also the channels labeled as bad FC5 and FC6

When you run `df =pd.DataFrame(data=reject_log.labels)` and calling the dataframe(df) you get:

0       | 1   | 2    | 3    | 4    | 5    | 6      | 7    | 8    | 9
--      | --  | --   | --   | --   | --   | --     | --   | --   | --
NaN | 1.0 | 1.0 | 2.0 | 1.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
... | ... | ... | ... | ... | ... | ... | ... | ... | ...
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0
NaN | 0.0 | 0.0 | 0.0 | 0.0 | 0.0 | NaN | 0.0 | 0.0 | 0.0

this shows that auto-reject indeed ignores the bad channels and does not interpolate them.

you can also see the indexes of the bad channels here:
   When you run `reject_log.ch_names`, you get 
   ['FC5', 'FC1', 'C3', 'CP5', 'CP1', 'FC2', 'FC6', 'C4', 'CP2', 'CP6']
      0         1       2      3       4       5         6       7     8         9       <--- index

@georgebuzzell  I plot before and after using auto-reject and it looked the same, it also shows both channels in white since they are marked as bad.

**Comment by georgebuzzell** [comment]
@SDOsmany this is great! Thank you for confirming this!!

@F-said as @SDOsmany as said, autoreject will interpolate the channels that it marks as bad, but WILL NOT interpolate the channels that we already marked as bad based on the badchans feature. THis is fine, and probably prerred. 

Here is what I suggest for moving forward. We switch to using autoreject for this feature, allowing it to interpolate as well. THen, we still need the feature-interp that @Jonhas is working on to interpolate the other channels that were previously marked as bad in our feature-badchans. So, @Jonhas , you should continue on with the interp feature as previously planned.

@SDOsmany how do you feel about trying to implement the autoreject feature here for this step, with help from @F-said ?

**Comment by SDOsmany** [comment]
@georgebuzzell sure, I will write something down and ask @F-said for feedback

**Comment by georgebuzzell** [comment]
@SDOsmany excellent! Thanks!!

**Comment by F-said** [comment]
@SDOsmany Nice! Looks great. Feel free to branch and implement this and I could do a code review when we merge it into the dev-feature branch.

**Comment by SDOsmany** [comment]
@georgebuzzell Hello George I have a question in your suggestion [here](https://ndclab.slack.com/archives/C01QX4D9VFT/p1620061708009300?thread_ts=1620061631.008600&cid=C01QX4D9VFT) Do we want to record how many epochs were dropped for each channel ? or you want the total epochs dropped for all channels? For example:
"FC5": 19 epochs dropped,
"C3": 17 epochs dropped.
or
"epochsDropped": 159 

Or maybe what you said to record was the number of times a channel was interpolated?
Sorry for the misunderstanding. 

this image represents my understanding of your suggestion
![Screenshot_5](https://user-images.githubusercontent.com/58539319/117079887-b2719b00-ad0a-11eb-8fa5-069d1ff7dabc.png)

**Comment by georgebuzzell** [comment]
@SDOsmany We would want to record how many epochs were dropped for each channel. So, just as you have it depicted in your image! Great; thanks! :)

**Comment by SDOsmany** [comment]
@F-said @SDOsmany finished up the final rejection feature. Please let me know if I need to change anything I would really appreciate some feedback.
here is the finalRej module

```python 
import mne
import autoreject as ar
import pandas as pd
import collections


def final_reject_epoch(user_params_dict, epochs, raw):
    """Final and automatic rejection of bad epochs
    @ preconditions: user parameters with the minimum and maximum for mne.Epoch() 
    @ preconditions: epochs specifically segments the raw dataset
    @ modifies: raw 
    @ postconditions: 
        - raw retains former dimensions  
            pre: (m x n) -- final_reject --> (m x n)
    """
    
    #creates the output dictionary to store the function output
    output_dict_finalRej = collections.defaultdict(dict)
    output_dict_finalRej['interpolatedChannels']=[]

    #fit epoch data using autoreject
    autoRej = ar.AutoReject()
    autoRej.fit(epochs)

    #Create a rejection log 
    reject_log = autoRej.get_reject_log(epochs)
    #Create a dataframe with the labels that indicate if a channel was interpolated
    df = pd.DataFrame(data=reject_log.labels,columns=raw.info.ch_names)
    for ch in raw.info.ch_names:
        if df[df[ch]==2][ch].count()>0:
            output_dict_finalRej['interpolatedChannels'].append(ch)

    # Get rejection dictionary to reject automatically detected bad epochs
    reject = ar.get_rejection_threshold(epochs)  
    
    for ch in raw.info.ch_names:
        rej_epochs = mne.Epochs(raw, epochs.events, event_id=epochs.event_id, tmin=user_params_dict['tmin'], tmax=user_params_dict['tmax'], #tmin=-0.2 tmax=0.5
                            picks=ch, preload=True, reject=reject, reject_by_annotation=True)
        
        # store amount of epochs dropped for each channel
        output_dict_finalRej['epochsDropped'][ch]=str(rej_epochs.drop_log.count((ch,)))
    return output_dict_finalRej
    ```
    
   and I was using this test for it
   
   ```python
   import mne
import autoreject as ar
import pandas as pd
import collections


def final_reject_epoch(user_params_dict, epochs, raw):
    """Final and automatic rejection of bad epochs
    @ preconditions: user parameters with the minimum and maximum for mne.Epoch() 
    @ preconditions: epochs specifically segments the raw dataset
    @ modifies: raw 
    @ postconditions: 
        - raw retains former dimensions  
            pre: (m x n) -- final_reject --> (m x n)
    """
    
    #creates the output dictionary to store the function output
    output_dict_finalRej = collections.defaultdict(dict)
    output_dict_finalRej['interpolatedChannels']=[]

    #fit epoch data using autoreject
    autoRej = ar.AutoReject()
    autoRej.fit(epochs)

    #Create a rejection log 
    reject_log = autoRej.get_reject_log(epochs)
    #Create a dataframe with the labels that indicate if a channel was interpolated
    df = pd.DataFrame(data=reject_log.labels,columns=raw.info.ch_names)
    for ch in raw.info.ch_names:
        if df[df[ch]==2][ch].count()>0:
            output_dict_finalRej['interpolatedChannels'].append(ch)

    # Get rejection dictionary to reject automatically detected bad epochs
    reject = ar.get_rejection_threshold(epochs)  
    
    for ch in raw.info.ch_names:
        rej_epochs = mne.Epochs(raw, epochs.events, event_id=epochs.event_id, tmin=user_params_dict['tmin'], tmax=user_params_dict['tmax'], #tmin=-0.2 tmax=0.5
                            picks=ch, preload=True, reject=reject, reject_by_annotation=True)
        
        # store amount of epochs dropped for each channel
        output_dict_finalRej['epochsDropped'][ch]=str(rej_epochs.drop_log.count((ch,)))
    return output_dict_finalRej
    ```

**Comment by SDOsmany** [comment]
@F-said I closed the last pull request and made a new one with the changes needed.

**Comment by F-said** [comment]
Closing per merged PR #100
